File: test/conformance/java/rows/BitsValuesValidData.java
// BitsValuesValidData — bits(N) fields: valid-data write/read acceptance
// (schema#876 schema matrix row java/BitsValuesValidData, the cell for the
// roadmap task bits-values/java/valid-data, contract §3.4 record, bits(N)).
//
// THE LAW (docs/FIXED-FORM-ALGORITHM.md:945; docs/SPEC-TABLES.md:10635,
// 16003). A `bits(N)` field's IMPLIED RANGE is `[0, 2^N - 1]` (the elided
// `min=0, max=2^N-1`). A value INSIDE that range, on both write and read,
// lands identically through the production encoder and decoder, with no
// counter moving: `clamped == 0`, `widened == 0`, `unknown == 0`,
// `kindMismatch == 0`, no refusal, no malformed. ACCEPTANCE means the
// round trip carries the value EXACTLY across the form-3 fixed wire.
//
// THE FIXED-FORM UNIT UNDER TEST. tables/examples/Ranges.schema declares
// `RangedWidths` — six `bits(N)` fields, two declared ranges that fit full
// storage (b8, b16, b32, b64) and two widths the column does NOT span
// (b12, b48). The unit is generated into build/tables-generated-java/examples
// (make/java.mk:88) and the conformance classpath (make/java.mk:585) carries
// `tabledemo.RangedWidthsFixed`, the form-3 read/write halves the contract
// points at.
//
// PRODUCTION PATH UNDER TEST (the actual call sites):
//   tabledemo.RangedWidthsFixed.save(values, count, buffer)  <- writer
//   tabledemo.RangedWidthsFixed.load(values, count, file,    <- reader
//                                    plan, remap, image, report)
// Both call sites reach the production runtime (`TableFixed.put32`,
// `TableFixed.get32`, `TableFixed.get64`, `RangedWidthsFixed.scatter`); no
// helper is the witness. The reader is driven end to end, through every
// reading check the form holds: `readHeader` (the framing), `select` (the
// hash taken as given against R.own_hash), the layout byte comparison, the
// prefill, and the scatter over the body.
//
// THE VECTOR. testdata/conformance/tables has no form-3 data for RangedWidths,
// so the bytes are constructed here from the law through the PRODUCTION WRITER
// — six fields, each filled with a UNIQUE VALID value, distinct from the
// declared default of 0 and from one another, so a misrouted field change is
// visible.
//
// RUN, FROM THE REPO ROOT:
//   java -cp build/conformance-java test/conformance/java/rows/BitsValuesValidData.java
// Exit 0 green, exit 1 red, one printed line per assertion.

import tabledemo.RangedWidthsFixed;
import tabledemo.TableFixed;

public final class BitsValuesValidData {
    private BitsValuesValidData() {}

    private static int failures = 0;

    private static void check(boolean ok, String what) {
        System.out.println((ok ? "ok   " : "RED  ") + what);
        if (!ok) { failures++; }
    }

    public static void main(String[] args) {
        // ---- the writer side ----------------------------------------------
        //
        // Six fields, six UNIQUE values, every one inside bits(N)'s implied
        // range [0, 2^N - 1]:
        //   b8  = 0xAB                (max 0xFF)
        //   b16 = 0xCAFE              (max 0xFFFF)
        //   b32 = 0xDEADBEEF          (a max-storable 32-bit pattern)
        //   b64 = 0xDECAFBADDEADBEEF  (max 0xFFFFFFFFFFFFFFFF)
        //   b12 = 0xABC               (= 2748, max 0xFFF)
        //   b48 = 0xC0FFEEDEADBEEF    (max 0xFFFFFFFFFFFF)
        // Each value is also different from the declared default of 0, so a
        // field that failed to write leaves a wrong read-back below.
        final RangedWidthsFixed.Value written = new RangedWidthsFixed.Value();
        written.b8  = 0xAB;
        written.b16 = 0xCAFE;
        written.b32 = 0xDEADBEEF;
        written.b64 = 0xDECAFBADDEADBEEFL;
        written.b12 = 0xABC;
        written.b48 = 0xC0FFEEDEADBEEFL;

        final int need = RangedWidthsFixed.measure(1);
        final byte[] file = new byte[need];
        final int saved = RangedWidthsFixed.save(new RangedWidthsFixed.Value[] { written }, 1, file);
        check(saved == need,
                "the writer fills exactly `measure(N)` bytes: " + saved + " of " + need
                + " bytes (header " + RangedWidthsFixed.headerBytes + " + 1 record of "
                + RangedWidthsFixed.recordBytes + ", with bodyBytes=" + RangedWidthsFixed.bodyBytes + ")");

        // ---- the writer's frame (every byte the form's own writeHeader set) ---
        //
        // file[0] = form byte 3; file[1..7] reserved, zero;
        // file[8..15] the layout hash (R.own_hash, the compile-time constant);
        // file[16..19] the u32 layout length; file[20..] the layout.
        final int recordsAt = RangedWidthsFixed.headerBytes;
        check(file[0] == 3,
                "the writer's frame: form byte 3 at file[0], got 0x" + (file[0] & 0xFF));
        for (int i = 1; i <= 7; i++) {
            if (file[i] != 0) {
                System.out.println("RED   reserved byte file[" + i + "] is not zero (0x"
                        + (file[i] & 0xFF) + ")");
                failures++;
                break;
            }
        }
        if (failures == 0) {
            check(true, "the writer's frame: the seven reserved bytes file[1..7] are zero");
        }
        check(TableFixed.get64(file, TableFixed.hashAt) == RangedWidthsFixed.hash,
                "the hash at 8 IS the compile-time constant R.own_hash (0x"
                        + Long.toHexString(RangedWidthsFixed.hash) + "L), the reader's own");
        check(TableFixed.get32(file, TableFixed.fileHeaderBytes) == RangedWidthsFixed.layout.length,
                "the u32 layout length at 16 is the layout's own length: "
                + TableFixed.get32(file, TableFixed.fileHeaderBytes) + " == "
                + RangedWidthsFixed.layout.length);

        // ---- the reader side ----------------------------------------------
        //
        // The production load path reads through every documented gate; the
        // assertions below state it must come back as one record, refuse
        // nothing, malform nothing and move no counter.
        final RangedWidthsFixed.Value[] readBack = { new RangedWidthsFixed.Value() };
        final TableFixed.Report report = new TableFixed.Report();
        final int n = RangedWidthsFixed.load(readBack, 1, file, RangedWidthsFixed.identityPlan(),
                new short[1024], RangedWidthsFixed.image(), report);

        // The READ itself completed: n == 1 (one record back), no refusal,
        // no malformed data, no layout refusal name carried.
        check(n == 1, "the production load reads back exactly one record (n=" + n + ")");
        check(!report.refused,
                "no refusal, and the file's own hash IS the reader's own (refused=" + report.refused + ")");
        check(!report.malformed,
                "no malformed framing on the writer's bytes (malformed=" + report.malformed + ")");

        // The ACCEPTANCE counters, all of them: §4's ledger, each zero on
        // valid input (a value in [0, 2^N - 1] isAccepted by `clamped` is N's
        // own ceiling, and no widening fires because every bits(N) carries its
        // own storage width).
        check(report.clamped == 0,
                "no `clamped` counter moves: every value is inside bits(N)'s implied [0, 2^N - 1] "
                + "(clamped=" + report.clamped + ")");
        check(report.widened == 0,
                "no `widened` counter moves: every bits(N) carries its own storage width "
                + "(widened=" + report.widened + ")");
        check(report.unknown == 0 && report.kindMismatch == 0,
                "the layout is the reader's own: no `unknown` and no `kind_mismatch` "
                + "(unknown=" + report.unknown + ", kindMismatch=" + report.kindMismatch + ")");

        // ---- the VALUES round-trip ----------------------------------------
        //
        // Every one of the six `bits(N)` fields lands EXACTLY the value the
        // writer set, the byte equality the wire is supposed to carry across
        // save -> load. Each pair is tested at its declared storage width:
        // b8/b16/b32/b12 are u32 lanes, b64/b48 are u64 lanes (see
        // RangedWidthsFixed.writeBody at offsets 0, 4, 8, 20 and 12, 24).
        check(readBack[0].b8  == written.b8,
                "b8 lands what was written: 0x" + Integer.toHexString(written.b8)
                + " (got 0x" + Integer.toHexString(readBack[0].b8) + ")");
        check(readBack[0].b16 == written.b16,
                "b16 lands what was written: 0x" + Integer.toHexString(written.b16)
                + " (got 0x" + Integer.toHexString(readBack[0].b16) + ")");
        check(readBack[0].b32 == written.b32,
                "b32 lands what was written: 0x" + Integer.toHexString(written.b32)
                + " (got 0x" + Integer.toHexString(readBack[0].b32) + ")");
        check(readBack[0].b64 == written.b64,
                "b64 lands what was written: 0x" + Long.toHexString(written.b64)
                + " (got 0x" + Long.toHexString(readBack[0].b64) + ")");
        check(readBack[0].b12 == written.b12,
                "b12 lands what was written: 0x" + Integer.toHexString(written.b12)
                + " (got 0x" + Integer.toHexString(readBack[0].b12) + ")");
        check(readBack[0].b48 == written.b48,
                "b48 lands what was written: 0x" + Long.toHexString(written.b48)
                + " (got 0x" + Long.toHexString(readBack[0].b48) + ")");

        System.out.println(failures == 0
                ? "BitsValuesValidData green: bits(N) fields accept valid data on the Java leg "
                        + "(write/read round-trip, no counter moved)"
                : "BitsValuesValidData RED: " + failures + " assertion(s) failed");
        if (failures != 0) {
            System.exit(1);
        }
    }
}

File: test/conformance/cpp/rows/BitsValuesValidData.cpp
// CELL cpp/BitsValuesValidData — bits(N) fields: valid-data write/read acceptance.
//
// THE LAW (docs/SPEC-TABLES.md §3.4, docs/FIXED-FORM-ALGORITHM.md:173): a
// `bits(N)` field rides at its DECLARED STORAGE WIDTH — 4 bytes for N <= 32
// and 8 bytes for N > 32 — and its implied range is [0, 2^N - 1]. A value
// past the top clamps and counts; a value at the cap is valid and moves
// nothing.
//
// WHAT THIS TEST DOES: builds one RangedWidths record (form 3, the fixed
// wire) with every bits(N) field set to its CAP (2^N - 1), writes it through
// the generated RangedWidthsFixedSave, reads it back through
// RangedWidthsFixedLoad, and asserts that every field survives the round trip
// with no counter moved. The same record is then written with a mid-range
// value (not 0, not max) to prove the path is not a constant fold.
//
// THE NEGATIVE CONTROL (`-DBITS_CONTROL`): flips ONE byte of the b12 field
// on the wire after the write, which lifts the decoded value past the cap
// (4095). The reader's bounds pass must clamp it back, so `clamped == 1` and
// `b12 == 4095`. A reader that ignores the implied range would pass the
// corrupted value through and go RED here.
//
// BUILD/RUN (from ./repo):
//   c++ -std=c++17 -Wall <conformance -I flags> build/tables-generated/examples/RangesTable.cpp
//     test/conformance/cpp/rows/BitsValuesValidData.cpp -o build/rows-cpp-BitsValuesValidData &&
//     ./build/rows-cpp-BitsValuesValidData
// Exit 0 green / exit 1 red, one printed line per assertion.
#include <cstdint>
#include <cstdio>
#include <cstring>
#include <vector>

#include "RangesTable.h"

static int failures = 0;

static void check( bool ok, const char * what )
{
    std::printf( "%s %s\n", ok ? "PASS" : "FAIL", what );
    if ( !ok ) failures++;
}

int main()
{
    // ---- proof: the body size matches the sum of storage widths ----
    // b8=4, b16=4, b32=4, b64=8, b12=4, b48=8 → 32 bytes
    check( tabledemo::RangedWidthsFixedBodyBytes == 32,
           "BitsValuesValidData: the body is 32 bytes" );

    // ---- CASE 1: every bits(N) field AT its cap (2^N - 1) ----
    {
        tabledemo::RangedWidths v;
        tabledemo::RangedWidthsReset( v );
        v.b8  = 0xFFu;              // 2^8  - 1
        v.b16 = 0xFFFFu;            // 2^16 - 1
        v.b32 = 0xFFFFFFFFu;        // 2^32 - 1
        v.b64 = 0xFFFFFFFFFFFFFFFFull; // 2^64 - 1
        v.b12 = 0xFFFu;             // 2^12 - 1 = 4095
        v.b48 = 0xFFFFFFFFFFFFull;  // 2^48 - 1

        std::vector<uint8_t> wire( (size_t) tabledemo::RangedWidthsFixedMeasure( 1 ) );
        const int64_t wrote = tabledemo::RangedWidthsFixedSave( &v, 1, wire.data(), (int64_t) wire.size() );
        check( wrote == (int64_t) wire.size(),
               "BitsValuesValidData cap: the record saves whole" );

        tabledemo::RangedWidths back;
        std::memset( (void *) &back, 0xAB, sizeof( back ) );
        tabledemo::TableReport r;
        const int64_t n = tabledemo::RangedWidthsFixedLoad(
            &back, 1, wire.data(), (int64_t) wire.size(),
            NULL, 0, NULL, &r );

        check( n == 1 && !r.refused && !r.malformed,
               "BitsValuesValidData cap: the record reads" );
        check( r.clamped == 0 && r.unknown == 0 && r.widened == 0,
               "BitsValuesValidData cap: at the cap moves no counter" );

        check( back.b8  == 0xFFu,              "BitsValuesValidData cap: b8  round-trips" );
        check( back.b16 == 0xFFFFu,            "BitsValuesValidData cap: b16 round-trips" );
        check( back.b32 == 0xFFFFFFFFu,        "BitsValuesValidData cap: b32 round-trips" );
        check( back.b64 == 0xFFFFFFFFFFFFFFFFull,
                                               "BitsValuesValidData cap: b64 round-trips" );
        check( back.b12 == 0xFFFu,             "BitsValuesValidData cap: b12 round-trips" );
        check( back.b48 == 0xFFFFFFFFFFFFull,  "BitsValuesValidData cap: b48 round-trips" );
    }

    // ---- CASE 2: mid-range values (not 0, not max) ----
    {
        tabledemo::RangedWidths v;
        tabledemo::RangedWidthsReset( v );
        v.b8  = 0xAAu;
        v.b16 = 0xBBBBu;
        v.b32 = 0xCCCCCCCCu;
        v.b64 = 0xDDDDDDDDDDDDDDDDull;
        v.b12 = 0xABCu;
        v.b48 = 0xEEEEEEEEEEEEull;

        std::vector<uint8_t> wire( (size_t) tabledemo::RangedWidthsFixedMeasure( 1 ) );
        const int64_t wrote = tabledemo::RangedWidthsFixedSave( &v, 1, wire.data(), (int64_t) wire.size() );
        check( wrote == (int64_t) wire.size(),
               "BitsValuesValidData mid: the record saves whole" );

        tabledemo::RangedWidths back;
        std::memset( (void *) &back, 0xAB, sizeof( back ) );
        tabledemo::TableReport r;
        const int64_t n = tabledemo::RangedWidthsFixedLoad(
            &back, 1, wire.data(), (int64_t) wire.size(),
            NULL, 0, NULL, &r );

        check( n == 1 && !r.refused && !r.malformed,
               "BitsValuesValidData mid: the record reads" );
        check( r.clamped == 0 && r.unknown == 0,
               "BitsValuesValidData mid: mid-range moves no counter" );

        check( back.b8  == 0xAAu,              "BitsValuesValidData mid: b8  round-trips" );
        check( back.b16 == 0xBBBBu,            "BitsValuesValidData mid: b16 round-trips" );
        check( back.b32 == 0xCCCCCCCCu,        "BitsValuesValidData mid: b32 round-trips" );
        check( back.b64 == 0xDDDDDDDDDDDDDDDDull,
                                               "BitsValuesValidData mid: b64 round-trips" );
        check( back.b12 == 0xABCu,             "BitsValuesValidData mid: b12 round-trips" );
        check( back.b48 == 0xEEEEEEEEEEEEull,  "BitsValuesValidData mid: b48 round-trips" );
    }

    // ---- CASE 3: the negative control — one forged byte lifts b12 past its cap ----
    {
        tabledemo::RangedWidths v;
        tabledemo::RangedWidthsReset( v );
        v.b8  = 0x42u;
        v.b16 = 0x4242u;
        v.b32 = 0x42424242u;
        v.b64 = 0x4242424242424242ull;
        v.b12 = 0x100u;             // a value inside the uint16 storage but past bits(12) cap
        v.b48 = 0x424242424242ull;

        std::vector<uint8_t> wire( (size_t) tabledemo::RangedWidthsFixedMeasure( 1 ) );
        const int64_t wrote = tabledemo::RangedWidthsFixedSave( &v, 1, wire.data(), (int64_t) wire.size() );
        check( wrote == (int64_t) wire.size(),
               "BitsValuesValidData over-cap: the record saves whole (the write already clamps)" );

        // The fixed writer clamps on the write side too, so the wire already
        // carries the clamped value. For the negative control, forge a byte
        // on the wire after the save to push b12 past 4095.
        // The layout: form(1) + hash(8) + layout-len(4) + layout + record-hash(8) + body(32).
        // b12 sits at offset 20 in the body (b8@0=4b, b16@4=4b, b32@8=4b, b12@20=4b).
        const size_t body_off = (size_t) tabledemo::kTableFixedHeaderBytes + 4
                              + (size_t) tabledemo::RangedWidthsFixedLayoutBytes + 8;
        // b12 is a uint32 at body+20; poke the MSB to make the value > 4095
        // but within the uint32 range the writer stored.
        // The writer stored 0x00000100 (clamped from 0x100 → 0x100 is 256,
        // which is within 0..4095 so it wasn't clamped by the writer).
        // Let's set it to 0x00002000 (8192, well past the 4095 cap).
        // On little-endian: byte 22 is the high byte of the upper 16 bits
        // of the 32-bit word. The value 0x00002000 stored LE is
        // [00 20 00 00]. We want to set byte body_off+22 to 0x20.
        wire[ body_off + 22 ] = 0x20; // was 0x00; now the wire carries 0x00002000

        tabledemo::RangedWidths back;
        std::memset( (void *) &back, 0xAB, sizeof( back ) );
        tabledemo::TableReport r;
        const int64_t n = tabledemo::RangedWidthsFixedLoad(
            &back, 1, wire.data(), (int64_t) wire.size(),
            NULL, 0, NULL, &r );

        check( n == 1 && !r.refused && !r.malformed,
               "BitsValuesValidData forged: the record reads (forged value is not a refusal)" );
        check( r.clamped == 1,
               "BitsValuesValidData forged: the b12 field clamps once" );
        check( back.b12 == 0xFFFu,
               "BitsValuesValidData forged: b12 clamps to 4095 (2^12 - 1)" );

        // neighbours stand
        check( back.b8 == 0x42u && back.b16 == 0x4242u && back.b32 == 0x42424242u,
               "BitsValuesValidData forged: the neighbour fields stand" );
    }

    return failures == 0 ? 0 : 1;
}
